### ISSUE TITLE
Validate if a Function is payable when sending a transaction

### ISSUE THREAD

**Issue by juanfranblanco** [Nethereum/Nethereum#379]
This should be applicable to ContractMessages only for backwards compatibility. 
The Function attribute will be decorated with payable.
If the amount is set at the time of creating a transaction input and/or sending/signing the transaction an exception will be thrown to either change the attribute or usage of an invalid function.

**Comment by juanfranblanco** [comment]
We should validate also views and pure transactions.

**Comment by TooPlain** [comment]
I could work on this, after reviewing function abi model and the json deserializer I believe this should be a quick add.
Does anyone know if there is any contrib guide for this repo?

It's better to start from somewhere then no where. Once that's done, I'll take my time on the contractmessages.

**Comment by juanfranblanco** [comment]
Thanks!! Although the original thought was to only do this for contract messages for backwards compatibility (mainly support past contracts..), but also this has been kept on hold as it may cause issues on Functions that are reused across with the same signature which might be payable or not, meaning that should be responsibility of the end user to understand if a function should be payable for a specific contract or not, regardless of the generic interface. Forcing an error will reduce flexibility on reusability of the "contract" layer, but also stop the necessary check by the developer on how to use a specific contract.

**Comment by TooPlain** [comment]
Ah alright please excuse my unfamiliarity with the layout of the source code, Can you point me to where contract messages is? So I may take a look. What I would have done with functionabi model and abi json deserializer was to just add a stateMutability string to the function abi struct and have the end user use the string to figure out if a function was payable, view, or nonPayable. In context of getting a contract that might've not been created by the end user, but I kinda of understand where your coming from.

**Comment by TooPlain** [comment]
Honestly this api has been great for me and is pretty powerful, I wanted to work on the docs once my application is out of beta as I have found that the current docs on the api are outdated or at least to me feel outdated.

**Comment by juanfranblanco** [comment]
> Ah alright please excuse my unfamiliarity with the layout of the source code, Can you point me to where contract messages is? So I may take a look. What I would have done with functionabi model and abi json deserializer was to just add a stateMutability string to the function abi struct and have the end user use the string to figure out if a function was payable, view, or nonPayable. In context of getting a contract that might've not been created by the end user, but I kinda of understand where your coming from.

Yes, this is my issue, as it may confuse a user in which lets say a function might be payable for a contract and not for another with the same interface. So keep it agnostic wont harm as it will be reused, some times the same function could be payable or not, as an example Mint in a ERC721 might accept some Ether but that does not mint will be payable all the time, and follow the same interface.

**Comment by juanfranblanco** [comment]
> 



> Honestly this api has been great for me and is pretty powerful, I wanted to work on the docs once my application is out of beta as I have found that the current docs on the api are outdated or at least to me feel outdated.

Yes!! documentation needs updated, i am leaning on the playground at the moment as the main source of both up to date examples and documentation. Any updates there will be truly helpful.
